### PR TITLE
Refactor with UDAs and mixin templates

### PR DESCRIPTION
**Pull request by iK4tsu**
```md
EventType
 * EventType is now a struct instead of an enum
 * EventType is called with an UDA [1]

basicEventType
 * mixin template to generate default Event functions [2]

genCallbacks
 * mixin template to generate callback functions and variables [3]
```
[1]
```d
@EventType("MyEvent")
class MyEvent : Event
{
    mixin basicEventType!MyEvent;
}
```
[2]
```d
@EventType("MyEvent")
class MyEvent : Event
{
    mixin basicEventType!MyEvent;

    // generates this
public:
    pragma(inline, true) @safe
    static @property string staticEventType()
    {
        import std.traits : getUDAs;
	return "MyEvent";
    }

    pragma(inline, true) @safe
    override @property string eventType() const
    {
        return staticEventType;
    }
}
```
[3]
```d
@EventType("MyFirstEvent")
class MyFirstEvent : Event
{
    mixin basicEventType!MyFirstEvent;
}

@EventType("MySecondEvent")
class MySecondEvent : Event
{
    mixin basicEventType!MySecondEvent;
public:
    this(in int x, in int y)
    {
        this.x = x;
        this.y = y;
    }
    const int x, y;
}
class Foo
{
    mixin genCallback!(Foo, MyFirstEvent) onMyFirst;
    // generates this
    // this functions are contained inside the scope onMyFirst
public:
    void connect(void delegate(in Foo, in MyFirstEvent) dg)
        in(dg.ptr !is null)
    {
        callback = dg;
    }

private:
    void dispach(in MyFirstEvent event)
    {
        if (callback.ptr !is null)
            callback(this, event);
    }

    void delegate(in Foo, in MyFirstEvent) callback;

/* ************************************************************************************************************ */

   mixin genCallback!(Foo, MySecondEvent, const int, "x", const int, "y") onMySecond;
    // generates this
    // this functions are contained inside the scope onMySecond
public:
    void connect(void delegate(in Foo, in int x, in int y) dg)
        in(dg.ptr !is null)
    {
        callback = dg;
    }

private:
    void dispach(in MySecondEvent event)
    {
        if (callback.ptr !is null)
            callback(this, event.x, event.y);
    }

    void delegate(in Foo, in int x, in int y) callback;

/* ************************************************************************************************************ */

    // create your functions to fire the signals
public:
    void myFirst()
    {
        // logic related to myFirst
        // emit the signal
        auto event = scoped!MyFirstEvent;
        onEvent(event);
    }

    void mySecond(in int x, in int y)
    {
        // logic related to mySecond
        // emit the signal
        auto event = scoped!MySecondEvent(x, y);
        onEvent(event);
    }
    
    // create an onEvent to dispach the events
private:
    void onEvent(in Event event)
    {
        auto ed = scoped!EventDispacher(event);
        ed.dispach!MyFirstEvent(&onMyFirst.dispach);
        ed.dispach!MySecondEvent(&onMySecond.dispach);
    } 
}

/* ************************************************************************************************************ */

// how to use
void onMyFirstEvent(in Foo sender, in MyFirstEvent event)
{
    // do stuf here
}

void onMySecondEvent(in Foo sender, in int x, in int y)
{
    // do stuf here
}

void  main()
{
    Foo foo = scoped!Foo;
    foo.myirst() // NO callback!
    foo.mySecond(4, 5); // NO callback!

    foo.onMyFirst.connect(toDelegate(&onMyFirstEvent));
    foo.onMySecond.connect(toDelegate(&onMySecondEvent));

    foo.myirst() // callback!
    foo.mySecond(4, 5); // callback!
}
```

### CODE REVIEW

File: source/event.d [hunk 1]
```diff
@@ -1,51 +1,360 @@
...
+ *
+ * Note: You need to declare an EventType UDA otherwise this won't compile!
+ *
+ * Params:
+ *     T = class type extended from Event
+ */
+mixin template basicEventType(T : Event)
 {
 public:
-	pragma(inline, true)
```
Comment (ljmf00) [inline source/event.d:30]: I do understand what you want here. You may got this from C++, but I do thing `pragma(inline, true)` should only be used if you really have to, because you may end up having heavier binaries if you use it in a wrong way.

I do prefer the compiler choice but this is up to you, but be aware of the side effects of this.

File: source/event.d [hunk 2]
```diff
@@ -1,51 +1,360 @@
...
+			this.y = y;
+		}
+
+		const size_t x, y;
 	}
+
+	import std.typecons : scoped;
+	class Foo
+	{
+		mixin genCallback!(const Foo, FooEvent, const int, "a", const int, "b", const int, "c") onFoo;
```
Comment (ljmf00) [inline source/event.d:298]: I guess you have this changed on other PR. If you do so, ignore this comment, but:

You can generate `onFoo` symbol inside the `mixin` by parsing the Event type name.

**Review by ljmf00 — APPROVED** [review]

**Comment by ljmf00** [comment]
Despites of the comments I left, the code seems very decent.

**Review by ljmf00 — CHANGES_REQUESTED** [review]
I miss clicked approve button. Please check my comments above.

File: source/event.d (hunk 2 above)
Comment (iK4tsu) [reply source/event.d:298]: No, `onFoo` is refering to to the mixin's scope. The mixin doesn't generate unique function names, every callback will have a `connect`, `dispach` and a `callback`delegate, because of this, you need to diferentiate between callbackA and callbackB since the functions have identical names and you can do that by naming the mixin's scope.

File: source/event.d [hunk 3]
```diff
@@ -1,51 +1,360 @@
...
+}
+
+@("Event: toEventFormat") @safe
+unittest
+{
+	assertEquals(toEventFormat!("a", "b"), (",event.a,event.b"));
+}
+
+
+version(unittest) import aurorafw.unit;
```
[the remark this answers is not in the document]
Comment (iK4tsu) [reply source/event.d:248]: Noted, I guess you have a point here, I'll change it in the last PR since this doesn't break anything.

File: source/event.d (hunk 1 above)
Comment (iK4tsu) [reply source/event.d:30]: But the function will always have the same behaviour, the compiler by default does not inline functions unless you tell it to.

Comment (ljmf00) [reply source/event.d:30]: > But the function will always have the same behaviour, the compiler by default does not inline functions unless you tell it to.

What? No, you are completely wrong. Let me deeply explain what inline does in D and in general what is the side effects of this:

Inlining, in general, should be a thing made by the compiler, in the optimization process.

The D spec says:
> - The default inline behavior is typically selectable with a compiler switch such as `-inline`.
> - Whether a particular function can be inlined or not is implementation defined.

"typically" doesn't mean "always". In practice, the DMD (default D compiler that properly implements D spec) doesn't inline in the optimization process, in general. See, here https://godbolt.org/z/qGMWMo, I ran `dmd` with `-O` for optimized code and also `-release` to suppress some debug behaviors, and even with `pragma(inline, true)` the function is not inlined.

On other hand, LDC compiler (llvm compiler implementation for D), without any flag (default behavior), it does inline when `pragma(inline, true)` is in action ( https://godbolt.org/z/K9EMrc ), although, if you leave the default behavior (removing `pragma(inline, true)`) it doesn't do inlining ( https://godbolt.org/z/vPW6s5 ), unless you tell to optimize it ( https://godbolt.org/z/nTdeaz ).

TL;DR

The compile DOES inline in some cases, unless you force (inline true) or force not to (inline false).

See, here, I don't force inlining and the function is inlined by the compiler on `-O3`: https://godbolt.org/z/q95KG9

Forcing inlining could lead to some side effects like heavy binaries when inlining really big functions when they are called several times and make debug process harder ( see the bottom of this article: http://sourceware.org/gdb/current/onlinedocs/gdb/Inline-Functions.html ).

Conclusion: As I discussed earlier with you, don't do `pragma(inline, true)` or `pragma(inline, false)` unless you really need to or you really know what you are doing.

File: source/event.d (hunk 2 above)
Comment (ljmf00) [reply source/event.d:298]: Sure, you're right. Maybe it's more readable and flexible too.

File: source/event.d (hunk 1 above)
Comment (iK4tsu) [reply source/event.d:30]: Thank you for this detailed explanation. I thought `prama(inline)` worked diferently. I'll change this in the third PR since this doesn't break anything here.
